contracts/core/VolOracle.sol

```diff
@@ -0,0 +1,264 @@
+    Accumulator public accumulator;
+
+    /// @dev Stores the last oracle TWAP price
+    uint256 public lastPrice;
+
+    /***
+     * Events
+     */
+
+    event Commit(
+        uint16 count,
+        uint32 commitTimestamp,
+        uint96 mean,
+        uint112 m2,
+        uint256 newValue,
+        address committer
+    );
+
+    /**
+     * @notice Creates an volatility oracle for a pool
+     * @param _pool is the Uniswap v3 pool
+     * @param _baseCurrency is the currency to measure the volatility of
+     * @param _quoteCurrency is the currency to quote the volatility in
+     * @param _period is how often the oracle needs to be updated
+     */
+    constructor(
+        address _pool,
+        address _baseCurrency,
+        address _quoteCurrency,
+        uint32 _period
+    ) {
+        IUniswapV3Pool uniPool = IUniswapV3Pool(_pool);
+        address token0 = uniPool.token0();
+        address token1 = uniPool.token1();
+
+        require(_pool != address(0), "!_pool");
+        require(_baseCurrency != address(0), "!_baseCurrency");
+        require(_quoteCurrency != address(0), "!_quoteCurrency");
+        require(_period > 0, "!_period");
+
+        // Check that the base and quote currencies are part of the pool
+        if (_baseCurrency == token0) {
+            require(_quoteCurrency == token1, "quote needs to be token1");
+        } else if (_baseCurrency == token1) {
+            require(_quoteCurrency == token0, "quote needs to be token0");
+        } else {
+            revert("No matching token");
+        }
+
+        pool = _pool;
+        baseCurrency = _baseCurrency;
+        quoteCurrency = _quoteCurrency;
+        baseCurrencyDecimals = IERC20Detailed(_baseCurrency).decimals();
+        period = _period;
+
+        // 31536000 seconds in a year
+        // divided by the period duration
+        // For e.g. if period = 1 day = 86400 seconds
+        // It would be 31536000/86400 = 365 days.
+        annualizationConstant = Math.sqrt(uint256(31536000).div(_period));
+    }
+
+    /**
+     * @notice Commits an oracle update
+     */
+    function commit() external {
+        (uint32 commitTimestamp, uint32 gapFromPeriod) = secondsFromPeriod();
+        require(gapFromPeriod < commitPhaseDuration, "Not commit phase");
+
+        uint256 price = twap();
+        uint256 _lastPrice = lastPrice;
+        // Result of the division is 10**18, but we scale down to 10**10 so it fits into uint112
+        uint256 periodReturn =
+            _lastPrice > 0 ? wdiv(price, _lastPrice).div(10**10) : 0;
+
+        Accumulator storage accum = accumulator;
+
+        require(
+            block.timestamp >=
+                accum.lastTimestamp + period - commitPhaseDuration,
+            "Committed"
+        );
+
+        (uint256 newCount, uint256 newMean, uint256 newM2) =
+            Welford.update(accum.count, accum.mean, accum.m2, periodReturn);
```

Comment:
so the new value in welford is the price t / price t-1?

Reply:
Yep, it's the return % of the period. So if ETH is 2000 at T-1, and goes to 3000 at T. The `periodReturn` is 1.5*10**8